examples/data-objects/inventory-app/src/view/inventoryList.tsx

```diff
@@ -3,34 +3,163 @@
+				{partViews}
+				<hr />
+				<button
+					onClick={() => root.parts.insertAtEnd(new Part({ name: "New Part", quantity: 0 }))}
+				>
+					Add Part
+				</button>
+			</div>
+		);
+	});
 
-	for (const part of inventory.parts) {
-		counters.push(
+/**
+ * A memoized auto-invalidated {@link Part}
+ */
+const PartView = withMemoizedTreeObservations(
+	({ part, remove }: { part: Part; remove: (part: Part) => void }) => (
+		<span>
 			<Counter
 				key={part.name}
 				title={part.name}
 				count={part.quantity}
 				onDecrement={(): number => part.quantity--}
 				onIncrement={(): number => part.quantity++}
-			/>,
+			/>
+			{/* Add an ability to remove the part.
+			This allows hitting a lot more edge cases
+			(like component unmount and cases where the React key in the array actually matters). */}
+			<button onClick={() => remove(part)}>Remove Part</button>
+		</span>
+	),
+);
+
+//
+// Below here are some unused examples of other ways to use the react tree utilities:
+//
+
+/**
+ * Example of a view which directly consumes multiple nodes from the tree.
+ */
+export const InventoryViewMonolithic =
+	// This could use a more granular observation strategy, like having a component for each part, but such an approach is not required
+	withTreeObservations(({ root }: { root: Inventory }) => {
+		const counters: JSX.Element[] = root.parts.map((part) => (
+			<Counter
+				key={part.name}
+				title={part.name}
+				count={part.quantity}
+				onDecrement={(): number => part.quantity--}
+				onIncrement={(): number => part.quantity++}
+			/>
+		));
+
+		return (
+			<div>
+				<h1>Inventory:</h1>
+				{counters}
+			</div>
 		);
-	}
+	});
+
+/**
+ * Example of a view which consumes part of a tree, delegating some to sub-components.
+ *
+ * This version uses usePropTreeNode. See InventoryView for a version using withMemoizedTreeObservations.
+ */
+export const InventoryViewWithHook: React.FC<{ root: PropTreeNode<Inventory> }> = ({
+	root,
+}) => {
+	const data = usePropTreeNode(root, (inventory: Inventory) => {
+		const partsList = inventory.parts;
+		// Example manually wrapping in PropNodes, showing how types without automatic support can still be made type safe.
+		// partsList.map((node) => toPropTreeNode(node)),
+		// Note that Array support is built in now, so this can just be:
+		const nodes = [...partsList];
+
+		// React's linter does not allow hooks in callbacks, but it is safe to suppress this for usePropTreeNode since it runs the callback immediately.
+		// eslint-disable-next-line react-hooks/rules-of-hooks
+		const removeChild = React.useCallback(
+			(part: Part) => partsList.removeAt(Tree.key(part) as number),
+			[partsList],
+		);
+
+		return { nodes, removeChild };
+	});
+
+	// Since usePropTreeNode is a hook, we can't use it on each item in this array.
+	// We can however use a component which uses the hook internally.
+	// Passing the node to the components as a PropTreeNode in its Props (as is done here)
```

Comment:
We don't wrap in PropTreeNode anymore right? That code got commented out in the hook above.

Reply:
This was intended to still use it: usePropTreeNode automatically wraps TreeNodes in its output. However the addition of `removeChild` causes that to stop happening, which still builds since PartView can accept either a `PropTreeNode<Part>` or a `Part`.

Let me see if I can tweak WrapNodes to work in more cases without being lossy on non tree node things.

Reply:
I did battle with TypeScript typing, and made WrapNodes far more complicated but hopefully much nicer, safer and general. It now wraps this case.

I also added explicit types for data above to make this clear (and stop compiling if it breaks)